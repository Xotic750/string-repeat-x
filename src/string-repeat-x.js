import numberIsFinite from 'is-finite-x';
import toInteger from 'to-integer-x';
import requireObjectCoercibleToString from 'require-coercible-to-string-x';
import attempt from 'attempt-x';

const EMPTY_STRING = '';
const {repeat: nativeRepeat} = EMPTY_STRING;
const hasNative = attempt(() => nativeRepeat.call('a', 5)).value === 'aaaaa';

/**
 * Repeat the given string the specified number of times.
 *
 * @param {string} value - The string to repeat.
 * @param {(number|string)} count - The number of times to repeat the string.
 * @returns {string} Repeated string.
 */
let $repeat;

if (hasNative) {
  $repeat = function repeat(value, count) {
    const result = nativeRepeat.call(requireObjectCoercibleToString(value), count);

    return typeof result === 'string' ? result : EMPTY_STRING;
  };
} else {
  $repeat = function repeat(value, count) {
    let string = requireObjectCoercibleToString(value);
    let n = toInteger(count);

    // Account for out-of-bounds indices
    if (n < 0 || !numberIsFinite(n)) {
      throw new RangeError('Invalid count value');
    }

    let result = EMPTY_STRING;
    while (n) {
      if (n % 2 === 1) {
        result += string;
      }

      if (n > 1) {
        string += string;
      }

      /* eslint-disable-next-line no-bitwise */
      n >>= 1;
    }

    return result;
  };
}

const r = $repeat;

export default r;
