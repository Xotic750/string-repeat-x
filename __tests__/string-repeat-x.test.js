import $repeat, {implementation} from '../src/string-repeat-x';

[implementation, $repeat].forEach((repeat, testNum) => {
  describe(`repeat ${testNum}`, function() {
    it('should return an empty string when a number is not given:', function() {
      expect.assertions(1);
      expect(repeat('a')).toBe('');
    });

    it('should return an empty string when zero or null is given as the number:', function() {
      expect.assertions(4);
      expect(repeat('', 0)).toBe('');
      expect(repeat('a', 0)).toBe('');
      expect(repeat('', null)).toBe('');
      expect(repeat('a', null)).toBe('');
    });

    it('should repeat the given string n times', function() {
      expect.assertions(11);
      expect(repeat(' ', 0)).toBe('');
      expect(repeat('a', 0)).toBe('');
      expect(repeat('a', 1)).toBe('a');
      expect(repeat('a', 2)).toBe('aa');
      expect(repeat('a', 3)).toBe('aaa');
      expect(repeat('   ', 3)).toBe('         ');
      expect(repeat('a ', 3)).toBe('a a a ');
      expect(repeat('a', 10)).toBe('aaaaaaaaaa');
      expect(repeat('b ', 10)).toBe('b b b b b b b b b b ');
      expect(repeat('a ', 10)).toBe('a a a a a a a a a a ');
      expect(repeat('abc ', 25)).toBe(
        'abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc ',
      );
    });

    it('should allow the multiplier to be a string:', function() {
      expect.assertions(6);
      expect(repeat('a', '0')).toBe('');
      expect(repeat('a', '1')).toBe('a');
      expect(repeat('a', '2')).toBe('aa');
      expect(repeat('a', '10')).toBe('aaaaaaaaaa');
      expect(repeat('b ', '10')).toBe('b b b b b b b b b b ');
      expect(repeat('a ', '10')).toBe('a a a a a a a a a a ');
    });

    it('should cache strings until the string changes:', function() {
      expect.assertions(6);
      expect(repeat('a', '5')).toBe('aaaaa');
      expect(repeat('b ', '5')).toBe('b b b b b ');
      expect(repeat('a ', '5')).toBe('a a a a a ');
      expect(repeat('c ', '5')).toBe('c c c c c ');
      expect(repeat('a ', '5')).toBe('a a a a a ');
      expect(repeat('b ', '5')).toBe('b b b b b ');
    });

    it('should throw an error when string is null or undefined:', function() {
      expect.assertions(3);
      expect(function() {
        repeat();
      }).toThrowErrorMatchingSnapshot();
      expect(function() {
        repeat(undefined);
      }).toThrowErrorMatchingSnapshot();
      expect(function() {
        repeat(null);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw an error when count is invalid:', function() {
      expect.assertions(2);
      expect(function() {
        repeat('a', -1);
      }).toThrowErrorMatchingSnapshot();
      expect(function() {
        repeat('a', Infinity);
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
