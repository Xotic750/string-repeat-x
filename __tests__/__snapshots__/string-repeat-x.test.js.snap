// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repeat 0 should throw an error when count is invalid: 1`] = `"Invalid count value"`;

exports[`repeat 0 should throw an error when count is invalid: 2`] = `"Invalid count value"`;

exports[`repeat 0 should throw an error when string is null or undefined: 1`] = `"Cannot call method on undefined"`;

exports[`repeat 0 should throw an error when string is null or undefined: 2`] = `"Cannot call method on undefined"`;

exports[`repeat 0 should throw an error when string is null or undefined: 3`] = `"Cannot call method on null"`;

exports[`repeat 1 should throw an error when count is invalid: 1`] = `"Invalid count value"`;

exports[`repeat 1 should throw an error when count is invalid: 2`] = `"Invalid count value"`;

exports[`repeat 1 should throw an error when string is null or undefined: 1`] = `"Cannot call method on undefined"`;

exports[`repeat 1 should throw an error when string is null or undefined: 2`] = `"Cannot call method on undefined"`;

exports[`repeat 1 should throw an error when string is null or undefined: 3`] = `"Cannot call method on null"`;
